### PR TITLE
Parse basic Dep3 headers (take two)

### PR DESCRIPTION
**Pull request by FauxFaux**
This is a re-up of #29.

I've addressed the review issues listed in the first pull request, and bought the tests up to the same style as the current tests.

I still believe it's a step in the right direction, if not totally sufficient to claim useful Dep3 behaviour.

I'm not convinced that the edit to `patch_series.py`'s `_parse_line` can be the only place that the Dep3 logic makes sense, but I reviewed all the other calls to Patch and they didn't seem relevant.

### CODE REVIEW

**Comment by agx** [comment]
This is moving the right way. Could you

- squash cleanups to previous things in the same series
- split out cleanups to other code in a separat commit (parts of modernise test syntax)
- Update the docs as mentioned in https://github.com/agx/git-buildpackage/pull/29

Thanks a lot for picking this up!

**Comment by FauxFaux** [comment]
Done.

I don't understand what you mean by removing the other cleanups from the "modernise tests" commit: the only change to other files there was to move a new utility method into a shared test class, to make the `repo` object make more sense. This is much more visible now that the two commits are squashed together.

As github apparently doesn't maintain old commits when squashing things on a pull request, I've archived the old pull request at https://github.com/FauxFaux/git-buildpackage/commits/pq_import_dep3-1.

File: gbp/patch_series.py [hunk 1]
```diff
@@ -117,6 +123,170 @@ def _get_subject_from_filename(self):
...
+    @property
+    def email(self):
+        """The patch author's email address"""
+        return self._get_info_field('email')
+
+    @property
+    def date(self):
+        """The patch's modification time"""
+        return self._get_info_field('date')
+
```
Comment (agx) [inline gbp/patch_series.py]: def email, def date move up here they need to be removed below. Othewise we end up with the identical methods on Patch and Dep3Patch

File: docs/chapters/patches.xml [hunk 2]
```diff
@@ -52,6 +52,15 @@
...
   </para>
 
+  <para>
+    Also, beware that &gbp-pq; currently has incomplete support for
+    <ulink url="http://dep.debian.net/deps/dep3/">DEP3</ulink> headers,
+    such as <computeroutput>Origin</computeroutput> and
+    <computeroutput>Forwarded</computeroutput>. These will be replaced
+    with <command>git-format-patch(1)</command> compatible options during
+    export.
+  </para>
```
Comment (agx) [inline docs/chapters/patches.xml]: As far as I read the patch 'author' and 'description' are handled separately while others might be dropped. Isn't this the case? Can you add a testcase for Origin and Forwarded?

File: gbp/patch_series.py (hunk 1 above)
Comment (FauxFaux) [reply gbp/patch_series.py]: `def author` and `def subject` too? This appears to have been a merge error, which I missed as it doesn't break any tests; good spot.

**Comment by agx** [comment]
Hi,
On Sun, Dec 17, 2017 at 12:11:02PM +0000, Chris West wrote:
> FauxFaux commented on this pull request.
> 
> 
> 
> > +
> +    @property
> +    def author(self):
> +        """The patch's author"""
> +        return self._get_info_field('author')
> +
> +    @property
> +    def email(self):
> +        """The patch author's email address"""
> +        return self._get_info_field('email')
> +
> +    @property
> +    def date(self):
> +        """The patch's modification time"""
> +        return self._get_info_field('date')
> +
> 
> `def author` and `def subject` too? This appears to have been a merge error, which I missed as it doesn't break any tests; good spot.

Yes all of them please (that's why I put it below the last one but it's
not obvious in githubs patch review GUI).

Thanks!

**Comment by FauxFaux** [comment]
I've added a test that shows what the new code does, squashed into the other test commit as that seems to be what we're going for:

https://github.com/agx/git-buildpackage/pull/37/commits/79f6e42721cdc78f269f7ed980dd6f4478c0c9fb#diff-03b4d4f6cf9441ebe4bb8d1fc7c15aa0R139

**Comment by ebourg** [comment]
Dumb question, would it be feasible not only to parse DEP3 headers but also to generate them? Thus using gbp pq with DEP3 patches would leave them unmodified.

**Comment by agx** [comment]
@ebourg sure that would be possible but someone would have to come around with patches for that. A one-way importer makes way more sense imho since having a git compatible format helps upstreams as well.

**Comment by agx** [comment]
@FauxFaux Thanks for the additional patch. This is helpful but we need this in the docs too. Somthing like the 2nd and 3rd paragraph of  75f847025eb348c4705a6f681a62bb14613e360e 's description.

**Comment by FauxFaux** [comment]
I've rephrased the docs to explain exactly what is happening:

> Also, beware that gbp pq currently has incomplete support for DEP3 headers. Initially, parsing with git-mailinfo(1) is attempted, which supports only the From and Subject fields. If neither of these are present, gbp pq will attempt to convert the patch from DEP3 format into a git-mailinfo(1) compatible format. This involves first loading From using the Author field and Subject using the first line of the Description field. Then, any additional fields (such as Origin and Forwarded), and the remainder of the Description (if any), will be appended to the body.

**Comment by agx** [comment]
Pushed now. @maxyz @FauxFaux thanks!

**Pull request closed without merging**